### ISSUE TITLE
fetch-pack: unexpected disconnect while reading sideband packet

### ISSUE THREAD

**Issue by roundspecs** [pingbird/puro#82]
I tried this on mac and linux. With different internet connections
```
$ puro create my_env 3.16.3 --log-level=3
[V] Started waiting for lock on /home/roundspecs/.puro/prefs.json
[V] Waiting for lock on /home/roundspecs/.puro/prefs.json took 0ms
[V] Checking if update is available
[V] Started waiting for lock on /home/roundspecs/.puro/latest_version
[V] Waiting for lock on /home/roundspecs/.puro/latest_version took 0ms
[V] Started waiting for lock on /home/roundspecs/.puro/releases_linux.json
[V] Waiting for lock on /home/roundspecs/.puro/releases_linux.json took 0ms
[V] Creating a new environment in /home/roundspecs/.puro/envs/my_env
[V] Started waiting for lock on /home/roundspecs/.puro/envs/my_env/update.lock
[V] Waiting for lock on /home/roundspecs/.puro/envs/my_env/update.lock took 0ms
[V] Started waiting for lock on /home/roundspecs/.puro/envs/my_env/prefs.json
[V] Waiting for lock on /home/roundspecs/.puro/envs/my_env/prefs.json took 0ms
[V] Pre-caching engine...
[V] Cloning flutter with shared refs
[V] Started cloning https://github.com/flutter/flutter.git
[V] Started checking if dart works
[V] Checking if dart works took 0ms
[W] dart version check failed, deleting cache
[V] Downloading engine
[V] Downloading https://storage.googleapis.com/flutter_infra_release/flutter/54a7145303f0dd9d0f93424a2e124eb4abef5091/dart-sdk-linux-x64.zip to /home/roundspecs/.puro/shared/caches/54a7145303f0dd9d0f93424a2e124eb4abef5091.zip
[V] Started downloading engine
[V] git failed with exit code 128
[E] git: Cloning into '/home/roundspecs/.puro/shared/flutter'...
    remote: Enumerating objects: 513947, done.
    remote: Counting objects:   5% (1/17)
    remote: Counting objects:  11% (2/17)
    remote: Counting objects:  17% (3/17)
    remote: Counting objects:  23% (4/17)
    remote: Counting objects:  29% (5/17)
    remote: Counting objects:  35% (6/17)
    remote: Counting objects:  41% (7/17)
    remote: Counting objects:  47% (8/17)
    remote: Counting objects:  52% (9/17)
    remote: Counting objects:  58% (10/17)
    remote: Counting objects:  64% (11/17)
    remote: Counting objects:  70% (12/17)
    remote: Counting objects:  76% (13/17)
    remote: Counting objects:  82% (14/17)
    remote: Counting objects:  88% (15/17)
    remote: Counting objects:  94% (16/17)
    remote: Counting objects: 100% (17/17)
    remote: Counting objects: 100% (17/17), done.
    remote: Compressing objects:   8% (1/12)
    remote: Compressing objects:  16% (2/12)
    remote: Compressing objects:  25% (3/12)
    remote: Compressing objects:  33% (4/12)
    remote: Compressing objects:  41% (5/12)
    remote: Compressing objects:  50% (6/12)
    remote: Compressing objects:  58% (7/12)
    remote: Compressing objects:  66% (8/12)
    remote: Compressing objects:  75% (9/12)
    remote: Compressing objects:  83% (10/12)
    remote: Compressing objects:  91% (11/12)
    remote: Compressing objects: 100% (12/12)
    remote: Compressing objects: 100% (12/12), done.
    Receiving objects:   0% (1/513947)
    Receiving objects:   0% (362/513947), 436.01 KiB | 419.00 KiB/s
    Receiving objects:   0% (2548/513947), 2.92 MiB | 168.00 KiB/s
    Receiving objects:   0% (2805/513947), 3.03 MiB | 174.00 KiB/s
    Receiving objects:   0% (3172/513947), 3.29 MiB | 193.00 KiB/s
    Receiving objects:   1% (5439/513947), 5.00 MiB | 75.00 KiB/s
    Receiving objects:   1% (5582/513947), 5.08 MiB | 79.00 KiB/s
    Receiving objects:   1% (5724/513947), 5.19 MiB | 88.00 KiB/s
    Receiving objects:   1% (5879/513947), 5.31 MiB | 98.00 KiB/s
    Receiving objects:   1% (6044/513947), 5.45 MiB | 108.00 KiB/s
    Receiving objects:   1% (6206/513947), 5.58 MiB | 119.00 KiB/s
    Receiving objects:   2% (12508/513947), 9.54 MiB | 74.00 KiB/s
    Receiving objects:   3% (17677/513947), 12.51 MiB | 107.00 KiB/s
    Receiving objects:   3% (17888/513947), 12.65 MiB | 125.00 KiB/s
    Receiving objects:   4% (20558/513947), 14.18 MiB | 185.00 KiB/s
    Receiving objects:   4% (20575/513947), 14.18 MiB | 185.00 KiB/s
    error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
    error: 6971 bytes of body are still expected
    fetch-pack: unexpected disconnect while reading sideband packet
    fatal: early EOF
    fatal: fetch-pack: invalid index-pack output
[V] Cloning https://github.com/flutter/flutter.git took 119156ms
[====           ] Downloading engine
[x] Bad state: git subprocess failed with exit code 128
    #0      GitClient._ensureSuccess (package:puro/src/git.dart:109)
    #1      GitClient.clone (package:puro/src/git.dart:167)
    <asynchronous suspension>
    #2      GitClient.cloneWithProgress.<anonymous closure> (package:puro/src/git.dart:181)
    <asynchronous suspension>
    #3      ProgressNode.wrap (package:puro/src/progress.dart:68)
    <asynchronous suspension>
    #4      GitClient.cloneWithProgress (package:puro/src/git.dart:179)
    <asynchronous suspension>
    #5      fetchOrCloneShared (package:puro/src/env/create.dart:205)
    <asynchronous suspension>
    #6      cloneFlutterWithSharedRefs (package:puro/src/env/create.dart:332)
    <asynchronous suspension>
    #7      createEnvironment.<anonymous closure> (package:puro/src/env/create.dart:148)
    <asynchronous suspension>
    #8      lockFile (package:puro/src/file_lock.dart:68)
    <asynchronous suspension>
    #9      createEnvironment (package:puro/src/env/create.dart:109)
    <asynchronous suspension>
    #10     CommandRunner.runCommand (package:args/command_runner.dart:212)
    <asynchronous suspension>
    #11     main (package:puro/src/cli.dart:256)
    <asynchronous suspension>

[x] Puro crashed! Please file an issue at https://github.com/pingbird/puro
    Consider running the command with a higher log level: `--log-level=4`
```

**Comment by roundspecs** [comment]
This fixed it
```sh
git config --global core.compression 0
```

**Comment by roundspecs** [comment]
This fixed it
```
git config --global core.compression 0
```

**Comment by burhanrashid52** [comment]
The way I found to fix this is to first use HTTP/1.1 and then increase the post buffer size.

Use HTTPS/1.1
```bash
git config --global http.version HTTP/1.1
```

And then increase the buffer size using
```bash
git config --global http.postBuffer 2038500000
```

However, the trick is to increase the buffer based on the bytes mentioned in the error message that says it cannot download. For example, the git error looks like this:

```bash
    Receiving objects:   5% (28560/532810), 19.44 MiB | 336.00 KiB/s
    error: RPC failed; curl 18 transfer closed with outstanding read data remaining
    error: 5319 bytes of body are still expected
    fetch-pack: unexpected disconnect while reading sideband packet
    fatal: early EOF
    fatal: fetch-pack: invalid index-pack output
```

Here, the line `error: 5319 bytes of body are still expected` mentions 5319 bytes. So, keep increasing the buffer size until it gets small enough, i.e., zero, to get this working.
```bash
git config --global http.postBuffer {additionalbytes}
```
